### ISSUE TITLE
FFI::NotFoundError: Function 'crypto_pwhash_memlimit_min' not found in [libsodium.so]

### ISSUE THREAD

**Issue by zedalaye** [crypto-rb/rbnacl#183]
I'm deploying on Debian 8 (Jessie) with latest updates and libsodium installed from jessie-backports, 

The version of libsodium that is installed is 1.0.11 (https://packages.debian.org/source/jessie-backports/libsodium)

I get this error when I try to require rbnacl :
`FFI::NotFoundError: Function 'crypto_pwhash_memlimit_min' not found in [libsodium.so]`

Do you know how to fix that ?

**Comment by tarcieri** [comment]
This seems like a regression in support for older versions of libsodium. Thanks for opening an issue.

Can you post a stack trace?

**Comment by zedalaye** [comment]
Like this one ?

```
FFI::NotFoundError: Function 'crypto_pwhash_memlimit_min' not found in [libsodium.so]
/home/myapp/shared/bundle/ruby/2.5.0/gems/ffi-1.9.25/lib/ffi/library.rb:273:in `attach_function'
/home/myapp/shared/bundle/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl/sodium.rb:38:in `sodium_constant'
/home/myapp/shared/bundle/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl/password_hash/argon2.rb:29:in `<class:Argon2>'
/home/myapp/shared/bundle/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl/password_hash/argon2.rb:12:in `<module:PasswordHash>'
/home/myapp/shared/bundle/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl/password_hash/argon2.rb:5:in `<module:RbNaCl>'
/home/myapp/shared/bundle/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl/password_hash/argon2.rb:4:in `<top (required)>'
/home/myapp/shared/bundle/ruby/2.5.0/gems/activesupport-5.2.0/lib/active_support/dependencies.rb:283:in `require'
/home/myapp/shared/bundle/ruby/2.5.0/gems/activesupport-5.2.0/lib/active_support/dependencies.rb:283:in `block in require'
/home/myapp/shared/bundle/ruby/2.5.0/gems/activesupport-5.2.0/lib/active_support/dependencies.rb:249:in `load_dependency'
/home/myapp/shared/bundle/ruby/2.5.0/gems/activesupport-5.2.0/lib/active_support/dependencies.rb:283:in `require'
/home/myapp/shared/bundle/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl.rb:78:in `<module:RbNaCl>'
/home/myapp/shared/bundle/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl.rb:24:in `<top (required)>'
/home/myapp/releases/20181119093721/config/application.rb:7:in `<top (required)>'
/home/myapp/releases/20181119093721/Rakefile:4:in `require_relative'
/home/myapp/releases/20181119093721/Rakefile:4:in `<top (required)>'
/home/myapp/shared/bundle/ruby/2.5.0/gems/rake-12.3.1/exe/rake:27:in `<top (required)>'
/home/deploy/.rbenv/versions/2.5.3/bin/bundle:23:in `load'
/home/deploy/.rbenv/versions/2.5.3/bin/bundle:23:in `<main>'
```

**Comment by tarcieri** [comment]
Yep, that's great, thanks. I will try to take a look at a fix when I have time.

If you're interested in doing a PR, the next steps are:

- Figure out which libsodium release added `crypto_pwhash_memlimit_min` (was it really that long after the other Argon2 functionality?)
- Gate the functionality on that libsodium release. It looks like there is something wrong with the gating here: https://github.com/crypto-rb/rbnacl/blob/e4d5811b5b75122fade3177a1196f696830309f7/lib/rbnacl/sodium/version.rb#L11

**Comment by tarcieri** [comment]
Hey @elijh, any ideas here?

**Comment by zedalaye** [comment]
>     * Figure out which libsodium release added `crypto_pwhash_memlimit_min` (it looks like it's supposed to be 1.0.13?)

According to libsodium release dates and crypto_pwhash.c history, this methods (and other ones that are related to limits) have been added in v1.0.12 (march 2017)

**Comment by tarcieri** [comment]
In theory the FFI code should not try to bind to them on libsodium versions less than `MINIMUM_LIBSODIUM_VERSION_FOR_ARGON2ID` (i.e. 1.0.13), so I'm not sure why it isn't gating it correctly.

**Comment by zedalaye** [comment]
https://github.com/crypto-rb/rbnacl/blob/157a6ab8042c9945baaa6916e855f67f8838d296/lib/rbnacl/password_hash.rb#L111-L118

It seems that Argon2 related methods are initialized since 1.0.9. We have 1.0.11 and Argon2 module requires memlimit and opslimit methods from libsodium that appeared only in 1.0.12

**Comment by tarcieri** [comment]
Yeah it looks like this symbol is mapped in on `1.0.9`+.

In the end it's just a numerical constant (`8192`, which is already even in the `.rb` file as a comment!), so one option is to avoid using FFI entirely by copying it into the Ruby code.

That won't automatically pick up upstream changes, but I don't expect this constant to change either.

**Comment by zedalaye** [comment]
Can't we add a check like `MINIMUM_LIBSODIUM_VERSION_FOR_LIMITS = [1, 0, 12].freeze` and initialize with default known values when `{mem,ops}limit` methods are not available ?

**Comment by tarcieri** [comment]
It's gating incorrectly on two different version numbers already. I'm not sure gating on a third version number is a good idea, especially when all it's doing is providing access to integer constants.

**Comment by prodoxx** [comment]
Have new commits been updated as a gem? I'm still having the same issue when I try to run my application with the latest rbnacl gem and latest libsodium build. I get the error:

FFI::NotFoundError: Function 'crypto_pwhash_memlimit_min' not found in [libsodium.so]
/cache/ruby/2.5.0/gems/ffi-1.10.0/lib/ffi/library.rb:273:in `attach_function'
/cache/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl/sodium.rb:38:in `sodium_constant'
/cache/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl/password_hash/argon2.rb:29:in `<class:Argon2>'
/cache/ruby/2.5.0/gems/rbnacl-6.0.0/lib/rbnacl/password_hash/argon2.rb:12:in `<module:PasswordHash>'

.......

**Comment by tarcieri** [comment]
Released as v6.0.1
